File: js/src/dieSimulator.ts
/*
 * @lc app=leetcode.cn id=1223 lang=typescript
 *
 * [1223] 掷骰子模拟
 */

// @lc code=start

const MOD = 1e9 + 7
let maxs: number[]
const dp: number[][][] = []

function dieSimulator(n: number, rollMax: number[]): number {
    maxs = rollMax
    for (let i = 0; i < n; ++i) {
        dp[i] = []
        for (let j = 0; j < 6; ++j) {
            dp[i].push([])
        }
    }
    return dfs(n, 0, rollMax[0])
}

function dfs(depth: number, prev: number, remain: number): number {
    if (depth-- == 0) return 1
    if (dp[depth][prev][remain] >= 0) return dp[depth][prev][remain]
    let ans = 0
    for (let i = 0; i < 6; ++i) {
        if (i != prev) ans += dfs(depth, i, maxs[i] - 1)
        else if (remain > 0) ans += dfs(depth, i, remain - 1)
    }
    return dp[depth][prev][remain] = ans % MOD
}

// @lc code=end


console.log(dieSimulator(2, [1, 1, 2, 2, 2, 3]))




`
    记忆化搜索
`

// const MOD = 1e9 + 7
// let maxs: number[]
// const dp: number[][][] = []

// function dieSimulator(n: number, rollMax: number[]): number {
//     maxs = rollMax
//     let ans = 0

//     dfs(n, 0, rollMax[0])

//     return ans
// }


// function dfs(depth: number, prev: number, remain: number): number {
//     if (depth === 0) return depth
//     if (dp[depth][prev][remain] > 0) return dp[depth][prev][remain]
//     let ans = 0

//     for (let i = 0; i < 6; ++i) {
//         if (i != prev) ans += dfs(depth - 1, i, maxs[i] - 1)
//         else if (remain > 0) ans += dfs(depth - 1, i, remain - 1)

//     }
//     return dp[depth][prev][remain] = ans % MOD

// }

File: js/src/DieSimulator.java
package day06;
/*
 * @lc app=leetcode.cn id=1223 lang=java [1223] 掷骰子模拟
 */

import java.util.Arrays;

// @lc code=start
class DieSimulator {
    long mod = 1_000_000_000 + 7;
    int[] rollMax = null;

    public int dieSimulator(int n, int[] rollMax) {
        int max = rollMax[0];
        for (int i = 1; i < 6; ++i) {
            max = max > rollMax[i] ? max : rollMax[i];
        }
        long[][][] dp = new long[n][6][max + 1];
        this.rollMax = rollMax;
        long sum = dfs(0, 0, 0, dp);
        return (int) (sum % mod);
    }

    private long dfs(int n, int prev, int times, long[][][] dp) {
        if (n < dp.length) {
            long tmp = 0;
            for (int i = 0; i < 6; ++i) {
                if (prev == i) {
                    if (times < rollMax[i]) {
                        if (dp[n][prev][times] != 0) tmp += dp[n][prev][times];
                        else tmp += dfs(n + 1, i, times + 1, dp);
                    }
                } else tmp += dfs(n + 1, i, 0, dp);
            }
            dp[n][prev][times] = tmp;
            return tmp;
        } else {
            return 1;
        }
    }
    // private
}
// @lc code=end
